File: packages/core/src/verify.ts
/**
 * Offline certificate-package verification (§6.3). Verification order:
 * resource bounds → parse/schema → key-role pins → object digests/signatures
 * → hash linkage → binding integrity → evidence validation → receipt
 * recomputation → anchoring → completeness.
 */

import { BondError, Head } from "./errors.js";
import { H, J, ZERO_HASH, deepEqual, parseJsonBytes, PACKAGE_LIMITS } from "./canon.js";
import {
  vSigned, vEntryBody, vReceiptBody, vAction, vUndoPlan, vPaperReview,
  checkPaperConsistency, vAssertion, vBlob,
} from "./schema.js";
import { verifySigned } from "./sign.js";
import { parseBlob, VerifyOptions, verifySource } from "./evidence.js";
import {
  RState, applyEntry, toView, ReducerEnv, MissingEvidence, residualEffect,
  remedyState, assemblyOf,
} from "./reducer.js";
import * as T from "./types.js";

const TRUTH = "ISSUER_ASSERTIONS_NOT_REAL_WORLD_PROOF";

function result(
  integrity: "VALID" | "INVALID",
  completeness: "COMPLETE" | "INCOMPLETE",
  currentness: "PINNED_PREFIX" | "UNANCHORED_PREFIX",
  head: Head | null,
  errors: string[],
): T.Verification {
  return {
    integrity, completeness, currentness,
    simulation: true, insurance: "PAPER_ONLY", truth: TRUTH, head, errors,
  };
}

/** Verify a certificate package offline. Never trusts issuer claims. */
export function verifyPackage(
  pkgBytes: Buffer,
  trust: T.TrustFile,
  expectedHead: Head | null,
  opts: VerifyOptions = {},
): T.Verification {
  const errors: string[] = [];
  try {
    if (pkgBytes.length > PACKAGE_LIMITS.maxBytes) throw new BondError("LIMIT", "package bytes");
    const pkgRaw = parseJsonBytes(pkgBytes, PACKAGE_LIMITS);
    if (typeof pkgRaw !== "object" || pkgRaw === null || Array.isArray(pkgRaw)) {
      throw new BondError("SCHEMA");
    }
    const o = pkgRaw as Record<string, unknown>;
    if (o.schema !== "bond.package/1") throw new BondError("UNSUPPORTED_VERSION");
    if (!Array.isArray(o.receipts) || !Array.isArray(o.entries) || !Array.isArray(o.blobs)) {
      throw new BondError("SCHEMA");
    }
    if (o.entries.length === 0) throw new BondError("SCHEMA", "empty package");
    if (o.entries.length > PACKAGE_LIMITS.maxEntries ||
      o.receipts.length > PACKAGE_LIMITS.maxReceipts ||
      o.blobs.length > PACKAGE_LIMITS.maxBlobs) {
      throw new BondError("LIMIT", "package member counts");
    }

    // Parse + validate all objects (closed schemas).
    const entries = o.entries.map((e) => vSigned(e, "entry", vEntryBody));
    const receipts = o.receipts.map((r) => vSigned(r, "receipt", vReceiptBody));
    const blobs = new Map<string, T.Blob>();
    for (const bRaw of o.blobs) {
      const b = vBlob(bRaw);
      if (b.data !== null && b.bytes > PACKAGE_LIMITS.maxBlobBytes) {
        throw new BondError("LIMIT", "blob bytes");
      }
      if (blobs.has(b.hash)) throw new BondError("SCHEMA", "duplicate blob");
      blobs.set(b.hash, b);
    }

    const b0 = entries[0]!.body;
    const tenantId = b0.tenant_id;
    const actionId = b0.action_id;
    const actionHash = b0.action_hash;

    // Signatures + hash linkage.
    let prevEntryHash = ZERO_HASH;
    for (let i = 0; i < entries.length; i++) {
      const e = entries[i]!;
      verifySigned("entry", e, trust, tenantId);
      const eb = e.body;
      if (eb.seq !== i + 1 || eb.previous_hash !== prevEntryHash ||
        eb.tenant_id !== tenantId || eb.action_id !== actionId || eb.action_hash !== actionHash) {
        throw new BondError("CHAIN_INVALID", "entry linkage");
      }
      prevEntryHash = e.hash;
    }
    if (receipts.length !== entries.length) {
      throw new BondError("PROJECTION_INVALID", "receipts do not cover entries");
    }
    let prevReceiptHash: string | null = null;
    for (let i = 0; i < receipts.length; i++) {
      const r = receipts[i]!;
      verifySigned("receipt", r, trust, tenantId);
      const rb = r.body;
      if (rb.revision !== i + 1 || rb.previous_receipt_hash !== prevReceiptHash ||
        rb.tenant_id !== tenantId || rb.action_id !== actionId || rb.action_hash !== actionHash ||
        rb.receipt_id !== receipts[0]!.body.receipt_id ||
        rb.head.seq !== i + 1 || rb.head.hash !== entries[i]!.hash) {
        throw new BondError("CHAIN_INVALID", "receipt linkage");
      }
      prevReceiptHash = r.hash;
    }
    const receiptId = receipts[0]!.body.receipt_id;
    const binding = receipts[0]!.body.binding;
    for (const r of receipts) {
      if (!deepEqual(r.body.binding, binding)) {
        throw new BondError("PROJECTION_INVALID", "binding changed mid-chain");
      }
    }

    // Binding integrity: resolve action/plan/paper blobs.
    let incomplete = false;
    const getBlob = (ref: T.ArtifactRef): T.Blob | null => {
      const b = blobs.get(ref.hash);
      if (!b || b.data === null || b.bytes !== ref.bytes || b.media_type !== ref.media_type) {
        return null;
      }
      return b;
    };
    const actionBlob = getBlob(binding.action);
    const planBlob = getBlob(binding.plan);
    const paperBlob = getBlob(binding.paper);
    if (!actionBlob || !planBlob || !paperBlob) {
      incomplete = true;
    }
    let action: T.Action | null = null;
    let plan: T.UndoPlan | null = null;
    if (actionBlob && planBlob) {
      action = vAction(parseJsonBytes(parseBlob(actionBlob), PACKAGE_LIMITS));
      plan = vUndoPlan(parseJsonBytes(parseBlob(planBlob), PACKAGE_LIMITS));
      if (action.action_id !== actionId || action.tenant_id !== tenantId ||
        action.execute_before !== undefined && false) {
        throw new BondError("PROJECTION_INVALID", "action binding mismatch");
      }
      if (plan.action_hash !== actionHash || plan.resource !== action.resource ||
        plan.adapter_build !== action.adapter_build ||
        plan.value_hash !== H(J(action.draft)) ||
        plan.expires_at !== action.terms.long_stop) {
        throw new BondError("PROJECTION_INVALID", "plan binding mismatch");
      }
    }
    if (paperBlob && action) {
      const paper = vSigned(parseJsonBytes(parseBlob(paperBlob), PACKAGE_LIMITS), "paper", vPaperReview);
      verifySigned("paper", paper, trust, tenantId);
      checkPaperConsistency(paper.body, action);
    }

    // Evidence validation + replay.
    const evCache = new Map<string, { facts: T.Facts; evidenceHash: string }>();
    const resolveEvidence = (ref: T.ArtifactRef): { facts: T.Facts; evidenceHash: string } | null => {
      const cached = evCache.get(ref.hash);
      if (cached) return cached;
      const blob = blobs.get(ref.hash);
      if (!blob || blob.data === null) return null;
      let bytes: Buffer;
      try {
        bytes = parseBlob(blob);
      } catch {
        throw new BondError("HASH_MISMATCH", "evidence blob");
      }
      const assertion = vSigned(parseJsonBytes(bytes, PACKAGE_LIMITS), "assertion", vAssertion);
      verifySigned("assertion", assertion, trust, tenantId);
      if (assertion.body.action_hash !== actionHash) {
        throw new BondError("PROJECTION_INVALID", "assertion action binding");
      }
      const srcBlob = blobs.get(assertion.body.source.hash);
      if (!srcBlob || srcBlob.data === null) {
        return null; // source bytes missing: incomplete
      }
      const srcBytes = parseBlob(srcBlob);
      verifySource(assertion.body.facts, assertion.body, srcBytes, trust, tenantId, opts);
      const out = { facts: assertion.body.facts, evidenceHash: ref.hash };
      evCache.set(ref.hash, out);
      return out;
    };

    const env: ReducerEnv | null = action && plan
      ? { action, plan, binding, resolve: resolveEvidence }
      : null;

    let st: RState | null = null;
    let replayOk = env !== null;
    for (let i = 0; i < entries.length; i++) {
      const entry = entries[i]!;
      if (env === null) break;
      try {
        const r = applyEntry(st, entry.body, entry.hash, env, "verify");
        st = r.state;
      } catch (e) {
        if (e instanceof MissingEvidence) {
          incomplete = true;
          replayOk = false;
          break;
        }
        throw e;
      }
      st.receiptId = receiptId;
      const rb = receipts[i]!.body;
      const derived = toView(st);
      if (!deepEqual(derived, rb.view)) {
        throw new BondError("PROJECTION_INVALID", `view mismatch at revision ${i + 1}`);
      }
      const derivedEvidence = st.evidenceRefs;
      if (!deepEqual(derivedEvidence, rb.evidence)) {
        throw new BondError("PROJECTION_INVALID", `evidence list mismatch at revision ${i + 1}`);
      }
      if (residualEffect(st) !== rb.residual_effect || remedyState(st) !== rb.remedy) {
        throw new BondError("PROJECTION_INVALID", `residual/remedy mismatch at revision ${i + 1}`);
      }
      if (assemblyOf(st, true) !== rb.assembly) {
        throw new BondError("PROJECTION_INVALID", `assembly mismatch at revision ${i + 1}`);
      }
    }
    if (!replayOk || incomplete) {
      if (incomplete && !errors.includes("INCOMPLETE")) errors.push("INCOMPLETE");
    }

    // Anchoring: a non-null expected head must equal the verified head
    // exactly; otherwise the pin fails closed with CHAIN_INVALID.
    const head: Head = {
      seq: entries[entries.length - 1]!.body.seq,
      hash: entries[entries.length - 1]!.hash,
    };
    let currentness: "PINNED_PREFIX" | "UNANCHORED_PREFIX" = "UNANCHORED_PREFIX";
    if (expectedHead !== null) {
      if (head === null || expectedHead.seq !== head.seq || expectedHead.hash !== head.hash) {
        throw new BondError("CHAIN_INVALID", "expected head does not anchor this prefix");
      }
      currentness = "PINNED_PREFIX";
    }

    const completeness: "COMPLETE" | "INCOMPLETE" =
      !incomplete && st !== null && assemblyOf(st, true) === "COMPLETE" &&
        receipts[receipts.length - 1]!.body.assembly === "COMPLETE"
        ? "COMPLETE"
        : "INCOMPLETE";

    return result("VALID", completeness, currentness, head, errors);
  } catch (e) {
    if (e instanceof BondError) {
      return result("INVALID", "INCOMPLETE", "UNANCHORED_PREFIX", null, [e.code]);
    }
    throw e;
  }
}

File: python/bond_verify/verify.py
"""Offline certificate-package verification (§6.3). Verification order:
resource bounds → parse/schema → key-role pins → object digests/signatures
→ hash linkage → binding integrity → evidence validation → receipt
recomputation → anchoring → completeness. Never trusts issuer claims."""

from __future__ import annotations

from .canon import (
    BondError, H, J, PACKAGE_LIMITS, ZERO_HASH, deep_equal, parse_json_bytes,
)
from .evidence import parse_blob, verify_source
from .reducer import (
    MissingEvidence, apply_entry, assembly_of, remedy_state, residual_effect,
    to_view,
)
from .schema import (
    check_paper_consistency, v_action, v_assertion, v_blob, v_entry_body,
    v_paper_review, v_receipt_body, v_signed, v_undo_plan,
)
from .sign import verify_signed

TRUTH = "ISSUER_ASSERTIONS_NOT_REAL_WORLD_PROOF"


def _result(integrity: str, completeness: str, currentness: str,
            head, errors: list) -> dict:
    return {
        "integrity": integrity,
        "completeness": completeness,
        "currentness": currentness,
        "simulation": True,
        "insurance": "PAPER_ONLY",
        "truth": TRUTH,
        "head": head,
        "errors": errors,
    }


def verify_package(pkg_bytes: bytes, trust: dict, expected_head=None,
                   source_verifiers: dict | None = None) -> dict:
    errors: list[str] = []
    try:
        if len(pkg_bytes) > PACKAGE_LIMITS.max_bytes:
            raise BondError("LIMIT", "package bytes")
        o = parse_json_bytes(pkg_bytes, PACKAGE_LIMITS)
        if not isinstance(o, dict):
            raise BondError("SCHEMA")
        if o.get("schema") != "bond.package/1":
            raise BondError("UNSUPPORTED_VERSION")
        if not isinstance(o.get("receipts"), list) \
                or not isinstance(o.get("entries"), list) \
                or not isinstance(o.get("blobs"), list):
            raise BondError("SCHEMA")
        if len(o["entries"]) == 0:
            raise BondError("SCHEMA", "empty package")
        if (len(o["entries"]) > PACKAGE_LIMITS.max_entries
                or len(o["receipts"]) > PACKAGE_LIMITS.max_receipts
                or len(o["blobs"]) > PACKAGE_LIMITS.max_blobs):
            raise BondError("LIMIT", "package member counts")

        entries = [v_signed(e, "entry", v_entry_body) for e in o["entries"]]
        receipts = [v_signed(r, "receipt", v_receipt_body) for r in o["receipts"]]
        blobs: dict[str, dict] = {}
        for b_raw in o["blobs"]:
            b = v_blob(b_raw)
            if b["data"] is not None and b["bytes"] > PACKAGE_LIMITS.max_blob_bytes:
                raise BondError("LIMIT", "blob bytes")
            if b["hash"] in blobs:
                raise BondError("SCHEMA", "duplicate blob")
            blobs[b["hash"]] = b

        b0 = entries[0]["body"]
        tenant_id = b0["tenant_id"]
        action_id = b0["action_id"]
        action_hash = b0["action_hash"]

        prev_entry_hash = ZERO_HASH
        for i, e in enumerate(entries):
            verify_signed("entry", e, trust, tenant_id)
            eb = e["body"]
            if (eb["seq"] != i + 1 or eb["previous_hash"] != prev_entry_hash
                    or eb["tenant_id"] != tenant_id or eb["action_id"] != action_id
                    or eb["action_hash"] != action_hash):
                raise BondError("CHAIN_INVALID", "entry linkage")
            prev_entry_hash = e["hash"]

        if len(receipts) != len(entries):
            raise BondError("PROJECTION_INVALID", "receipts do not cover entries")
        prev_receipt_hash = None
        for i, r in enumerate(receipts):
            verify_signed("receipt", r, trust, tenant_id)
            rb = r["body"]
            if (rb["revision"] != i + 1
                    or rb["previous_receipt_hash"] != prev_receipt_hash
                    or rb["tenant_id"] != tenant_id or rb["action_id"] != action_id
                    or rb["action_hash"] != action_hash
                    or rb["receipt_id"] != receipts[0]["body"]["receipt_id"]
                    or rb["head"]["seq"] != i + 1
                    or rb["head"]["hash"] != entries[i]["hash"]):
                raise BondError("CHAIN_INVALID", "receipt linkage")
            prev_receipt_hash = r["hash"]

        receipt_id = receipts[0]["body"]["receipt_id"]
        binding = receipts[0]["body"]["binding"]
        for r in receipts:
            if not deep_equal(r["body"]["binding"], binding):
                raise BondError("PROJECTION_INVALID", "binding changed mid-chain")

        # Binding integrity: resolve action/plan/paper blobs.
        incomplete = False

        def get_blob(ref):
            b = blobs.get(ref["hash"])
            if (not b or b["data"] is None or b["bytes"] != ref["bytes"]
                    or b["media_type"] != ref["media_type"]):
                return None
            return b

        action_blob = get_blob(binding["action"])
        plan_blob = get_blob(binding["plan"])
        paper_blob = get_blob(binding["paper"])
        if not action_blob or not plan_blob or not paper_blob:
            incomplete = True

        action = plan = None
        if action_blob and plan_blob:
            action = v_action(parse_json_bytes(parse_blob(action_blob), PACKAGE_LIMITS))
            plan = v_undo_plan(parse_json_bytes(parse_blob(plan_blob), PACKAGE_LIMITS))
            if action["action_id"] != action_id or action["tenant_id"] != tenant_id:
                raise BondError("PROJECTION_INVALID", "action binding mismatch")
            if (plan["action_hash"] != action_hash
                    or plan["resource"] != action["resource"]
                    or plan["adapter_build"] != action["adapter_build"]
                    or plan["value_hash"] != H(J(action["draft"]))
                    or plan["expires_at"] != action["terms"]["long_stop"]):
                raise BondError("PROJECTION_INVALID", "plan binding mismatch")
        if paper_blob and action:
            paper = v_signed(parse_json_bytes(parse_blob(paper_blob), PACKAGE_LIMITS),
                             "paper", v_paper_review)
            verify_signed("paper", paper, trust, tenant_id)
            check_paper_consistency(paper["body"], action)

        # Evidence validation + replay.
        ev_cache: dict[str, dict] = {}

        def resolve_evidence(ref):
            cached = ev_cache.get(ref["hash"])
            if cached:
                return cached
            blob = blobs.get(ref["hash"])
            if not blob or blob["data"] is None:
                return None
            try:
                data = parse_blob(blob)
            except BondError:
                raise BondError("HASH_MISMATCH", "evidence blob")
            assertion = v_signed(parse_json_bytes(data, PACKAGE_LIMITS),
                                 "assertion", v_assertion)
            verify_signed("assertion", assertion, trust, tenant_id)
            if assertion["body"]["action_hash"] != action_hash:
                raise BondError("PROJECTION_INVALID", "assertion action binding")
            src_blob = blobs.get(assertion["body"]["source"]["hash"])
            if not src_blob or src_blob["data"] is None:
                return None  # source bytes missing: incomplete
            src_bytes = parse_blob(src_blob)
            verify_source(assertion["body"]["facts"], assertion["body"],
                          src_bytes, trust, tenant_id, source_verifiers)
            out = {"facts": assertion["body"]["facts"], "evidenceHash": ref["hash"]}
            ev_cache[ref["hash"]] = out
            return out

        env = ({"action": action, "plan": plan, "binding": binding,
                "resolve": resolve_evidence}
               if action and plan else None)

        st = None
        replay_ok = env is not None
        for i, entry in enumerate(entries):
            if env is None:
                break
            try:
                st, _ = apply_entry(st, entry["body"], entry["hash"], env, "verify")
            except MissingEvidence:
                incomplete = True
                replay_ok = False
                break
            st["receiptId"] = receipt_id
            rb = receipts[i]["body"]
            if not deep_equal(to_view(st), rb["view"]):
                raise BondError("PROJECTION_INVALID",
                                f"view mismatch at revision {i + 1}")
            if not deep_equal(st["evidenceRefs"], rb["evidence"]):
                raise BondError("PROJECTION_INVALID",
                                f"evidence list mismatch at revision {i + 1}")
            if (residual_effect(st) != rb["residual_effect"]
                    or remedy_state(st) != rb["remedy"]):
                raise BondError("PROJECTION_INVALID",
                                f"residual/remedy mismatch at revision {i + 1}")
            if assembly_of(st, True) != rb["assembly"]:
                raise BondError("PROJECTION_INVALID",
                                f"assembly mismatch at revision {i + 1}")
        if incomplete and "INCOMPLETE" not in errors:
            errors.append("INCOMPLETE")

        # Anchoring: a non-null expected head must equal the verified head.
        head = {"seq": entries[-1]["body"]["seq"], "hash": entries[-1]["hash"]}
        currentness = "UNANCHORED_PREFIX"
        if expected_head is not None:
            if (expected_head["seq"] != head["seq"]
                    or expected_head["hash"] != head["hash"]):
                raise BondError("CHAIN_INVALID",
                                "expected head does not anchor this prefix")
            currentness = "PINNED_PREFIX"

        completeness = (
            "COMPLETE"
            if (not incomplete and st is not None
                and assembly_of(st, True) == "COMPLETE"
                and receipts[-1]["body"]["assembly"] == "COMPLETE")
            else "INCOMPLETE"
        )
        return _result("VALID", completeness, currentness, head, errors)
    except BondError as e:
        return _result("INVALID", "INCOMPLETE", "UNANCHORED_PREFIX", None, [e.code])
